### PR TITLE
TST Add Playwright for an auxiliary testing framework

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

_Updated:_

Adds auxiliary playwright runner for testing.

```
pytest --runner [playwright, selenium]
```

__

~~I proposed changing our testing framework from Selenium to Playwright (#2048). However, Playwright is not battle-tested in pyodide environment, and it has some performance issues in Firefox. Therefore, I think it is not ready to be the primary testing framework of pyodide for now.~~

~~Instead, this PR makes Playwright an auxiliary testing framework, so that it can be used together with Selenium.
I expect Playwright can help us especially when Selenium is unstable.~~

~~When `PLAYWRIGHT` env variable is set, pytest will use Playwright instead of Selenium.~~

- [x] Update documentation

### CODE REVIEW

**Comment by rth** [comment]
Thanks for working on this @ryanking13 and sorry for slow feedback. 

I think it's an interesting idea. I haven't looked in the code in detail yet, however a few general comments,
 - I think having a second `conftest_playwright.oy`, with significant code duplication is not ideal and is going to be difficult to maintain. Actually, even our conftest.py is getting quite large by now. So it could be a good time to,
   - factorize what we can into some new module in `pyodide_build`. We already have `pyodide_build.testing` but it would probably be better to put it in a separate file, as we don't want selenium to be a mandatory run time dependency of pyodide_build (optional dependency would be OK). 
   - For instance, maybe we could use a base class with methods that are re-used between SeleniumWrapper / PlaywrightWrapper. 
  
   This can be done in a separate PR.
  
-  If we do the previous work than maybe the `conftest.py` would be short enough to put both selenium and playwright configs there.
- If we are going to add it, it needs to run in automatically triggered CI otherwise there is a high likelihood that we would break it with future changes. I don't know where it would be better to put it. We could use it on GH Actions for Chrome for instance, if the setup is not too complex.
- For now, maybe let's not add it to the requirements.txt since it's an optional feature (but install in CI when needed).
- For using this runner, I think it would be better to add [a pytest option](https://docs.pytest.org/en/6.2.x/example/markers.html#custom-marker-and-command-line-option-to-control-test-runs) so that we could run it with `pytest --runner=playwright` (with the default being `selenium`)

**Comment by ryanking13** [comment]
> * I think having a second `conftest_playwright.oy`, with significant code duplication is not ideal and is going to be difficult to maintain. Actually, even our conftest.py is getting quite large by now. So it could be a good time to,

Yeah, I definitely agree. I didn't want to make `conftest.py` more complicated, but having an another configuration file looks worse.
I'll look into `pyodide-build` to see what we can do.


> * If we are going to add it, it needs to run in automatically triggered CI otherwise there is a high likelihood that we would break it with future changes. I don't know where it would be better to put it. We could use it on GH Actions for Chrome for instance, if the setup is not too complex.

If we are going to automatically trigger both Selenium and Playwright tests, I think GH Actions would be a good place to put it (Playwright only or maybe both). Acutally, I also have interests in migrating CircleCI jobs to GH Actions.

> * For now, maybe let's not add it to the requirements.txt since it's an optional feature (but install in CI when needed).

Agreed.

> * For using this runner, I think it would be better to add [a pytest option](https://docs.pytest.org/en/6.2.x/example/markers.html#custom-marker-and-command-line-option-to-control-test-runs) so that we could run it with `pytest --runner=playwright` (with the default being `selenium`)

Thank you for the suggestion! I didn't know that option.

**Comment by ryanking13** [comment]
Sorry for the delay. Here is an update, applying suggestions from @rth.
(It looks like the SQLite URL is temporarily dead, I'll re-trigger CI when it's restored)

1. I restructured the test wrapper class, adding a base class `BrowserWrapper`:

```
 ┏ BrowserWrapper
 ┣  SeleniumWrapper 
 ┃  ┣ SeleniumFirefoxWrapper
 ┃  ┗ SeleniumChromeWrapper
 ┣  PlaywrightWrapper 
 ┃  ┣ PlaywrightFirefoxWrapper
 ┃  ┗ PlaywrightChromeWrapper
 ┗  NodeWrapper
```

The `conftest.py` got much bigger due to this change. So in a subsequent PR, I would like to move some general testing-related stuff in a new module, as @rth mentioned.

2. Use playwright for GH actions

My first thought in this PR was to use Playwright for a 2nd testing framework for CI, where we can use it optionally when Selenium is unstable. But as @rth said, there is a high likelihood that we would break it with future changes if it is not automatically triggered. Therefore, I updated to use playwright in GH actions, so that we can double-test in both frameworks.
But actually, I think this is a temporary way and can be changed at any time. Maybe we can run every test in both frameworks later if our CI resource becomes affluent.
(BTW, It feels like selenium tests are quite stable after browser version update yay :) )

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -113,21 +115,26 @@ jobs:
         shell: bash -l {0}
         run: |
           pip install -r requirements.txt
+          # FIXME: playwright>=1.18.0 fails in testing `test_isinstance` due to the
+          #        'Maximum call stack size exceeded' error
+          pip install "playwright<1.18.0" && python -m playwright install
```
Comment (ryanking13) [inline .github/workflows/main.yml:120]: Playwright >= 1.18.0 (which comes with chrome >= v99) fails on this recursion test. I'm not sure this is a Playwright issue or a Chrome issue.

https://github.com/python/cpython/blob/0a7dcbdb13f1f2ab6e76e1cff47e80fb263f5da0/Lib/test/test_isinstance.py#L274-L282

**Comment by ryanking13** [comment]
Back in this PR, I think this is ready to be reviewed.

Compared to the time that this PR was originally suggested, our Selenium test became a lot stable. Still, I think adding support for playwright have some benefits such as:

- Webkit (Safari) support with no additional effort
- Easy to install for new comers

File: src/tests/test_pyproxy.py [hunk 2]
```diff
@@ -369,10 +369,10 @@ def test_pyproxy_get_buffer_type_argument(selenium, array_type):
         if fmt.lower() == "q":
             assert result == [hex(x).replace("0x", "") for x in list(mv.cast(fmt))]
         elif fmt == "f" or fmt == "d":
-            from math import isclose
+            from math import isclose, isnan
 
             for a, b in zip(result, list(mv.cast(fmt))):
-                if a and b:
+                if a and b and not (isnan(a) or isnan(b)):
```
Comment (hoodmane) [inline src/tests/test_pyproxy.py:375]: This test should probably be improved, the `f` and `d` cases are pretty sloppy...
If they are nan shouldn't both be nan?

**Review by hoodmane — APPROVED** [review]

File: src/tests/test_pyproxy.py (hunk 2 above)
Comment (ryanking13) [reply src/tests/test_pyproxy.py:375]: > If they are nan shouldn't both be nan?

Both should be NaN, but I am getting some strange results in selenium and node:

```py
@pytest.mark.skip_pyproxy_check
def test_temp(selenium):
    selenium.run_js(
        """
        let a = pyodide.runPython("bytes(range(252, 256))")
        let buf = a.getBuffer("f32")
        console.log(buf.data[0])
        """
    )
```

This should print `NaN`, but I get `None` in selenium and node. The playwright prints `NaN` correctly.
It also works correctly in the browsers, so maybe there is a bug in our selenium code.

**Comment by ryanking13** [comment]
Thanks for the reviews!

**Comment by hoodmane** [comment]
Thanks @ryanking13!

**Pull request merged**